network/retrieve/retrieve.go

```diff
@@ -0,0 +1,445 @@
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/metrics"
+	"github.com/ethereum/go-ethereum/node"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/p2p/enode"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/ethersphere/swarm/chunk"
+	"github.com/ethersphere/swarm/log"
+	"github.com/ethersphere/swarm/network"
+	"github.com/ethersphere/swarm/network/timeouts"
+	"github.com/ethersphere/swarm/p2p/protocols"
+	"github.com/ethersphere/swarm/spancontext"
+	"github.com/ethersphere/swarm/storage"
+	opentracing "github.com/opentracing/opentracing-go"
+	olog "github.com/opentracing/opentracing-go/log"
+)
+
+var (
+	// Compile time interface check
+	_ node.Service = &Retrieval{}
+
+	// Metrics
+	processReceivedChunksCount    = metrics.NewRegisteredCounter("network.retrieve.received_chunks.count", nil)
+	handleRetrieveRequestMsgCount = metrics.NewRegisteredCounter("network.retrieve.handle_retrieve_request_msg.count", nil)
+	retrieveChunkFail             = metrics.NewRegisteredCounter("network.retrieve.retrieve_chunks_fail.count", nil)
+
+	lastReceivedRetrieveChunksMsg = metrics.GetOrRegisterGauge("network.retrieve.received_chunks", nil)
+
+	// Protocol spec
+	spec = &protocols.Spec{
+		Name:       "bzz-retrieve",
+		Version:    1,
+		MaxMsgSize: 10 * 1024 * 1024,
+		Messages: []interface{}{
+			ChunkDelivery{},
+			RetrieveRequest{},
+		},
+	}
+
+	ErrNoPeerFound = errors.New("no peer found")
+)
+
+// Retrieval holds state and handles protocol messages for the `bzz-retrieve` protocol
+type Retrieval struct {
+	mtx      sync.Mutex
+	netStore *storage.NetStore
+	kad      *network.Kademlia
+	peers    map[enode.ID]*Peer //this protocols peers
+	spec     *protocols.Spec    //this protocol's spec
+	balance  protocols.Balance  //implements protocols.Balance, for accounting
+	prices   protocols.Prices   //implements protocols.Prices, provides prices to accounting
+
+	quit chan struct{} // termination
+}
+
+// NewRetrieval returns a new instance of the retrieval protocol handler
+func NewRetrieval(kad *network.Kademlia, ns *storage.NetStore) *Retrieval {
+	ret := &Retrieval{
+		kad:      kad,
+		peers:    make(map[enode.ID]*Peer),
+		netStore: ns,
+		quit:     make(chan struct{}),
+	}
+
+	ret.spec = spec
+
+	return ret
+}
+
+func (r *Retrieval) addPeer(p *Peer) {
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+	r.peers[p.ID()] = p
+}
+
+func (r *Retrieval) removePeer(p *Peer) {
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+	delete(r.peers, p.ID())
+}
+
+func (r *Retrieval) getPeer(id enode.ID) *Peer {
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+
+	return r.peers[id]
+}
+
+// Run protocol function
+func (r *Retrieval) Run(p *p2p.Peer, rw p2p.MsgReadWriter) error {
+	peer := protocols.NewPeer(p, rw, r.spec)
+	bp := network.NewBzzPeer(peer)
+	sp := NewPeer(bp)
+	r.addPeer(sp)
+	defer r.removePeer(sp)
+	return peer.Run(r.HandleMsg(sp))
+}
+
+// HandleMsg handles incoming messages for a certain peer
```

Comment:
this does not need to be exported and can be called `handlerForPeer` or just `getHandler`?